src/GafferScene/Instancer.cpp

```diff
@@ -403,6 +393,92 @@ class Instancer::EngineData : public Data
 					throw IECore::Exception( fmt::format( "Context primitive variable for \"{}\" is not a correctly sized Vertex primitive variable", v.name.string() ) );
 				}
 			}
+
+			if( !m_numValidPrototypes )
+			{
+				// We don't need to build m_instancesForPrototype if we're not outputting any prototypes
+				return;
+			}
+
+			int constantPrototypeIndex = -1;
+			if( !m_prototypeIndices )
+			{
+				constantPrototypeIndex = m_prototypeIndexRemap[ 0 ];
+				if( constantPrototypeIndex == -1 )
+				{
+					// If we have no indices to specify other prototypes, and the first prototype is
+					// invalid, we're not going to output anything, and can early exit
+					return;
+				}
+			}
+
+			std::vector< std::vector<int> > instancesForPrototypeIndex( m_numPrototypes );
+
+			if( constantPrototypeIndex == -1 || m_ids )
+			{
+				if( m_ids )
```

Comment:
You had a bunch of questions about this code and its performance, so I've ended up refactoring it and adding more perf tests: `test[ChildNames/EngineData]PerfWithPrototypes(AndIds)` ( and probably spending too much time rewriting in history in an attempt to make it make sense through various commits ).

I've removed the inner loop special case for "constantPrototypeIndex but with m_ids", but kept the one special case for "we're literally just setting this vector to an identity mapping for 0 ... N - 1", since that did have a measurable performance effect.